### ISSUE TITLE
package structure and API

### ISSUE THREAD

**Issue by DirkEilander** [Deltares/hydromt#959]
### Kind of request

Changing existing functionality

### Enhancement Description

I suggest to reorganize the package structure and API to make it more recognizable and easier to use for users and developers. 

At the highest level, the package structure should be organized around the main concepts of the package `DataCatalog` and `Model`, all submodules and scripts should in my opinion as much as possible be organized around these concepts in the `data_catalog` and `model` submodules. Additional submodules are `cli`, `gis`, `io`, `processes` (was workflows), `stats`, and `utils`.	 

**Suggested package structure:**

```bash
hydromt/
├── __init__.py # expose DataCatalog, Model
├── _compat.py
├── config.py
|── plugins.py  # TODO move to cli/ ?
├── _typing/
├── _utils/
│   ├── log.py # TODO this is used quite often in plugin examples, can be moved to utils?
├── _validators/
├── cli/
├── data_catalog/
│   ├── __init__.py  # expose DataCatalog, PredefinedCatalog
│   ├── predefined_catalogs.py
│   ├── data_catalog.py
|   ├── adapters/
|   |  ├── caching.py # TODO can be moved to _utils?
|   |  ├── utils.py # TODO has many unused functions; can be moved to _utils?
|   ├── sources/
|   ├── drivers/
|   ├── uri_resolvers/ 
├── gis/
├── io/
├── processes/ # NOTE: was workflows
├── model/ # NOTE: was models (plural)
│   ├── __init__.py  # expose Model, ModelComponent
│   ├── model.py
│   ├── root.py
│   ├── region.py  # NOTE: was workflows/region.py
│   ├── components/
├── stats/
├── utils/
```	

**remove scripts?**
hydromt/models/plugins.py -> this is superseded by plugins.py

### Use case

Easier structure for developers and users

### Additional Context

This also touches on some of the discussion in https://github.com/Deltares/hydromt/discussions/862

**Comment by hboisgon** [comment]
I would be in favor of this restructure. I think it makes it easier to navigate the repository and find the relevant scripts. I think the main reason why the models and data catalog and drivers were not put together was for the discovery of the components/drivers so re-organizing means solving the other issue you opened in #958 

I also think log.py is quite used in plugins so I would agree to move it to utils folder. Also good to rename the ``workflows`` folder to ``processes``.

**Comment by Jaapel** [comment]
Hi Dirk!
- Happy to move all things data under the umbrella of `data_catalog`. I think that makes sense.
- `io` should move to data_catalog in that case? I would see `io` as a module where we store shared methods to decode / encode data.
- `plugins.py` is generic over all the places where plugins can appear: (`Model`, `ModelComponent`, `Driver`, `MetaDataResolver` (soon to be `URIResolver` I guess?). Therefore it should be at the top level.
- The HydroMT `cli` should just be responsible for parsing arguments and running some code from the other modules. If there is functionality in the `cli` submodule that does things specifically for the CLI, I think that code should be elsewhere so we can reuse it in the future.

**Comment by savente93** [comment]
Closed by #974
